### PR TITLE
[Merged by Bors] - feat: add reverse induction principle for Vector

### PR DESCRIPTION
**Pull request by alexkeizer**
Add a snoc pseudo-constructor, lemmas to reason about snoc, and a reverse induction principle for Vectors.

---

`Vector.snoc xs x` is just a short-hand for `xs ++ [x]`.
The reverse induction principle uses `xs.snoc x` for it's recursive step, which makes it easier to reason about algorithms that work on a `Vector` right-to-left, such as `foldr` or `mapAccumr`.
The PR also add some simplification rules for `snoc` which help with such backwards proofs.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#4994
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4994~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Data/Vector/Snoc.lean [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+      This function has two arguments: `nil` handles the base case on `C nil`,
+      and `snoc` defines the inductive step using `∀ x : α, C xs → C (xs.snoc x)`.
+
+      This can be used as `induction v using Vector.revInductionOn`. -/
+  @[elab_as_elim]
+  def revInductionOn {C : ∀ {n : ℕ}, Vector α n → Sort _} {n : ℕ} (v : Vector α n)
+                      (nil : C nil)
+                      (snoc : ∀ {n : ℕ} (xs : Vector α n) (x : α), C xs → C (xs.snoc x))
+                      : C v :=
```
Comment (ChrisHughes24) [inline Mathlib/Data/Vector/Snoc.lean:86]: This indentation style is not usual. We just indent the theorem statement with four spaces usually, with occasional extra two space indentations for readability.

**Review by ChrisHughes24 — COMMENTED** [review]
I think my main comments here are just style comments about indentation.

**Comment by alexkeizer** [comment]
> I think my main comments here are just style comments about indentation.

These should now be addressed

File: Mathlib/Data/Vector/Snoc.lean [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  cases xs
+  simp only [reverse, cons, toList_mk, List.reverse_cons, snoc]
+  congr
+
+@[simp]
+theorem reverse_snoc : reverse (xs.snoc x) = x ::ᵥ (reverse xs) := by
+  cases xs
+  simp only [reverse, snoc, cons, toList_mk]
+  congr
+  simp[toList, (·++·), Vector.append, Append.append]
```
Comment (ChrisHughes24) [inline Mathlib/Data/Vector/Snoc.lean:51]:
```suggestion
  simp [toList, (·++·), Vector.append, Append.append]
```

File: Mathlib/Data/Vector/Snoc.lean [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+theorem replicate_succ_to_snoc (val : α) :
+    replicate (n+1) val = (replicate n val).snoc val := by
+  clear xs
+  induction n
+  case zero => rfl
+  case succ n ih =>
+    rw[replicate_succ]
+    conv => {
+      rhs; rw[replicate_succ]
```
Comment (ChrisHughes24) [inline Mathlib/Data/Vector/Snoc.lean:63]:
```suggestion
    rw [replicate_succ]
    conv => {
      rhs; rw [replicate_succ]
```

File: Mathlib/Data/Vector/Snoc.lean [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      cast (by simp) <| snoc xs.reverse ys.reverse x y r)
+
+/-- Define `C v` by *reverse* case analysis, i.e. by handling the cases `nil` and `xs.snoc x`
+    separately -/
+@[elab_as_elim]
+def revCasesOn {C : ∀ {n : ℕ}, Vector α n → Sort _} {n : ℕ} (v : Vector α n)
+    (nil : C nil)
+    (snoc : ∀ {n : ℕ} (xs : Vector α n) (x : α), C (xs.snoc x)) :
+    C v :=
+  revInductionOn v nil fun xs x _ => snoc xs x
```
Comment (ChrisHughes24) [inline Mathlib/Data/Vector/Snoc.lean:117]:
```suggestion
  revInductionOn v nil fun xs x _ => snoc xs x
  
```

File: Mathlib/Data/Vector/Snoc.lean [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+end Induction
+
+
```
Comment (ChrisHughes24) [inline Mathlib/Data/Vector/Snoc.lean:120]:
```suggestion
```

**Review by ChrisHughes24 — APPROVED** [review]
You can merge after the small fixes

bors d+

**Comment by alexkeizer** [comment]
bors merge

**Pull request closed without merging**